test/data/EnumerableSet.ts

```diff
@@ -0,0 +1,195 @@
+            ethers.constants.MaxUint256.toString(),
+          );
+        });
+      });
+
+      describe('#length()', () => {
+        it('returns length of enumerable set', async () => {
+          expect(await instance['length()']()).to.equal(0);
+
+          await instance['add(bytes32)'](zeroBytes32);
+          expect(await instance['length()']()).to.equal(1);
+
+          await instance['add(bytes32)'](oneBytes32);
+          expect(await instance['length()']()).to.equal(2);
+
+          await instance['add(bytes32)'](twoBytes32);
+          expect(await instance['length()']()).to.equal(3);
+
+          await instance['remove(bytes32)'](twoBytes32);
+          expect(await instance['length()']()).to.equal(2);
+
+          await instance['remove(bytes32)'](oneBytes32);
+          expect(await instance['length()']()).to.equal(1);
+
+          await instance['remove(bytes32)'](zeroBytes32);
+          expect(await instance['length()']()).to.equal(0);
+        });
+      });
+
+      describe('#add(bytes32)', () => {
+        it('adds value to set', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          await instance['add(bytes32)'](oneBytes32);
+          await instance['add(bytes32)'](twoBytes32);
+
+          expect(await instance['toArray()']()).to.deep.equal([
+            zeroBytes32,
+            oneBytes32,
+            twoBytes32,
+          ]);
+        });
+
+        it('returns true if value is added', async () => {
+          expect(await instance.callStatic['add(bytes32)'](zeroBytes32)).to.be
+            .true;
+        });
+
+        it('returns false if value has already been added', async () => {
+          await instance.callStatic['remove(bytes32)'](zeroBytes32);
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.false;
+        });
+      });
+
+      describe('#remove(bytes32)', () => {
+        it('removes value from set', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          await instance['add(bytes32)'](oneBytes32);
+          await instance['add(bytes32)'](twoBytes32);
+
+          await instance['remove(bytes32)'](zeroBytes32);
+          expect(await instance['toArray()']()).to.deep.equal([
+            twoBytes32,
+            oneBytes32,
+          ]);
+        });
+
+        it('returns true if value is removed', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          await instance['add(bytes32)'](oneBytes32);
+          await instance['add(bytes32)'](twoBytes32);
+
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.true;
+        });
+
+        it('returns false if value is not removed', async () => {
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.false;
+        });
+
+        it('removes value from index mapping and array', async () => {
```

Comment:
I would probably move these tests to the `length` and `indexOf` blocks.  Seems like too many things are happening here.  I don't feel strongly about it, however.

Reply:
I think the purpose is to test all effects of the function in an effort for self-containment.

The remove function explicitly alters these properties.

Reply:
If we separate test concerns perfectly, the `remove` tests must test the side effects to `length` and `indexOf`, and both the `length` and `indexOf` tests must test what happens when `remove` is called.  This is a bit too redundant, so we must choose which approach makes more sense.  There's no right answer.

Reply:
I removed the "removes value from index mapping and array" test case. The array length is already checked when values are added/ removed from the set in the "returns length of enumerable set" test case. I modified the test case "returns index of the value" to check that the mapping is updated when values are added/ removed from the set.

https://github.com/solidstate-network/solidstate-solidity/pull/144/commits/ffb9354f2cb3b9c6855cf7689266180c7c95d3c8

Reply:
I will also update this test case in the BinaryHeap PR, as its checked there as well.